File: Robot Control Course/Chapter12/Lesson5/Chapter12_Lesson5.java

import java.util.concurrent.*;
import java.util.*;
import java.lang.Math;

class JointInterface {
    public double readPosition() { /* ... */ return 0.0; }
    public double readVelocity() { /* ... */ return 0.0; }
    public void writeTorque(double tau) { /* ... */ }
}

public class RealTimeController {
    public static void main(String[] args) {
        final double Ts = 0.002; // 2 ms
        final double Kp = 50.0;
        final double Kd = 2.0;
        final double tFinal = 5.0;

        JointInterface joint = new JointInterface();

        List<Double> log_t = new ArrayList<>();
        List<Double> log_Tk = new ArrayList<>();
        List<Double> log_jitter = new ArrayList<>();
        List<Double> log_q = new ArrayList<>();
        List<Double> log_qd = new ArrayList<>();
        List<Double> log_e = new ArrayList<>();

        ScheduledExecutorService exec = Executors.newSingleThreadScheduledExecutor();

        final long startNano = System.nanoTime();
        final double[] ePrev = new double[]{0.0};
        final long periodNano = (long)(Ts * 1e9);

        Runnable task = new Runnable() {
            long lastNano = startNano;

            @Override
            public void run() {
                long nowNano = System.nanoTime();
                double t = (nowNano - startNano) / 1e9;
                if (t >= tFinal) {
                    exec.shutdown();
                    return;
                }

                double Tk = (nowNano - lastNano) / 1e9;
                lastNano = nowNano;
                double jitter = Tk - Ts;

                double q  = joint.readPosition();
                double dq = joint.readVelocity();

                double qd  = 0.5 * Math.sin(2.0 * Math.PI * 0.5 * t);
                double dqd = 0.5 * 2.0 * Math.PI * 0.5 * Math.cos(2.0 * Math.PI * 0.5 * t);

                double e  = qd - q;
                double de = (e - ePrev[0]) / Math.max(Tk, 1e-6);
                ePrev[0] = e;

                double tau = Kp * e + Kd * de;
                joint.writeTorque(tau);

                log_t.add(t);
                log_Tk.add(Tk);
                log_jitter.add(jitter);
                log_q.add(q);
                log_qd.add(qd);
                log_e.add(e);
            }
        };

        exec.scheduleAtFixedRate(task, 0, periodNano, TimeUnit.NANOSECONDS);
    }
}

File: Tutorials/Codes/Robot_Control_Course/Chapter12/Lesson5/Chapter12_Lesson5.m

Ts = 0.002;
Kp = 50;
Kd = 2;
tFinal = 5.0;

% Robot interface: implement these as needed
readJoint  = @() deal(0.0, 0.0);  % [q, dq]
writeTorque = @(tau) [];          % send torque

t0 = tic;
lastTime = 0.0;
nextTime = Ts;

e_prev = 0.0;

log_t = [];
log_Tk = [];
log_jitter = [];
log_q = [];
log_qd = [];
log_e = [];

while true
    now = toc(t0);
    if now >= tFinal
        break;
    end

    Tk = now - lastTime;
    lastTime = now;
    jitter = Tk - Ts;

    [q, dq] = readJoint();

    qd  = 0.5 * sin(2*pi*0.5*now);
    dqd = 0.5 * 2*pi*0.5 * cos(2*pi*0.5*now);

    e  = qd - q;
    de = (e - e_prev) / max(Tk, 1e-6);
    e_prev = e;

    tau = Kp * e + Kd * de;
    writeTorque(tau);

    log_t(end+1,1) = now;
    log_Tk(end+1,1) = Tk;
    log_jitter(end+1,1) = jitter;
    log_q(end+1,1) = q;
    log_qd(end+1,1) = qd;
    log_e(end+1,1) = e;

    % Sleep to enforce nominal Ts
    pause(max(0.0, nextTime - now));
    nextTime = nextTime + Ts;
end

% Plot log_Tk, log_jitter, and log_e vs log_t
